ethcore/wasm/src/env.rs

```diff
@@ -14,183 +14,276 @@
+		&[I32, I32, I32, I32],
+		Some(I32),
+	);
+
+	pub const SUICIDE: StaticSignature = StaticSignature(
 		&[I32],
 		None,
-	),
-	Static(
-		"_origin",
+	);
+
+	pub const BLOCKHASH: StaticSignature = StaticSignature(
+		&[I64, I32],
+		None,
+	);
+
+	pub const BLOCKNUMBER: StaticSignature = StaticSignature(
+		&[],
+		Some(I64),
+	);
+
+	pub const COINBASE: StaticSignature = StaticSignature(
 		&[I32],
 		None,
-	),
-	Static(
-		"_address",
+	);
+
+	pub const DIFFICULTY: StaticSignature = StaticSignature(
 		&[I32],
 		None,
-	),
-	Static(
-		"_value",
+	);
+
+	pub const GASLIMIT: StaticSignature = StaticSignature(
 		&[I32],
 		None,
-	),
-	Static(
-		"_timestamp",
-		&[],
-		Some(I64),
-	),
-	Static(
-		"_blocknumber",
+	);
+
+	pub const TIMESTAMP: StaticSignature = StaticSignature(
 		&[],
 		Some(I64),
-	),
-	Static(
-		"_difficulty",
+	);
+
+	pub const ADDRESS: StaticSignature = StaticSignature(
 		&[I32],
 		None,
-	),
-	Static(
-		"_gaslimit",
+	);
+
+	pub const SENDER: StaticSignature = StaticSignature(
 		&[I32],
 		None,
-	),
-	Static(
-		"_elog",
-		&[I32; 4],
-		None,
-	),
-
-	// TODO: Get rid of it also somehow?
-	Static(
-		"_llvm_trap",
-		&[I32; 0],
-		None
-	),
-
-	Static(
-		"_llvm_bswap_i64",
-		&[I64],
-		Some(I64)
-	),
-];
-
-pub fn native_bindings<'a>(runtime: &'a mut Runtime) -> interpreter::UserDefinedElements<'a, UserTrap> {
-	interpreter::UserDefinedElements {
-		executor: Some(runtime),
-		globals: ::std::collections::HashMap::new(),
-		functions: ::std::borrow::Cow::from(SIGNATURES),
+	);
+
+	pub const ORIGIN: StaticSignature = StaticSignature(
+		&[I32],
+		None,
+	);
+
+	pub const ELOG: StaticSignature = StaticSignature(
+		&[I32, I32, I32, I32],
+		None,
+	);
+
+	impl Into<wasmi::Signature> for StaticSignature {
+		fn into(self) -> wasmi::Signature {
+			wasmi::Signature::new(self.0, self.1)
+		}
+	}
+}
+
+fn host(signature: signatures::StaticSignature, idx: usize) -> FuncRef {
+	FuncInstance::alloc_host(signature.into(), idx)
+}
+
+/// Import resolver for wasmi
+/// Maps all functions that runtime support to the corresponding contract import
+/// entries.
+/// Also manages initial memory request from the runtime.
+#[derive(Default)]
+pub struct ImportResolver {
+	max_memory: u32,
+	memory: RefCell<Option<MemoryRef>>,
+}
+
+impl ImportResolver {
+	/// New import resolver with specifed maximum amount of inital memory.
+	pub fn with_limit(max_memory: u32) -> ImportResolver {
+		ImportResolver {
+			max_memory: max_memory,
+			.. Default::default()
+		}
+	}
+
+	/// Returns memory that was instantiated during the contract module
+	/// start. If contract does not use memory at all, the dummy memory of length (0, 0)
+	/// will be created instead. So this method always returns memory instance
+	/// unless errored.
+	pub fn memory_ref(&self) -> Result<MemoryRef, Error> {
+		{
+			let mut mem_ref = self.memory.borrow_mut();
+			if mem_ref.is_none() { *mem_ref = Some(MemoryInstance::alloc(0, Some(0))?); }
```

Comment:
This invocation actually can't return error since params `(0, Some(0))` are perfectly valid

Reply:
But i don't want to depend on this invariant from other lib

Reply:
But otherwise this code embeds this invariant into the consensus